Pausable chunked HTTP responses

DaveCTurner opened the pull request:
Today we must collect together in memory everything needed to send a
HTTP response before starting to send it to the client, even if we're
using the `chunked` transfer encoding to bound the memory needed for the
final serialization step.

To properly bound the memory usage on the coordinating node we must
instead be able to start sending the response to the client before we
have collected everything needed to finish it. If we do this then we
must be able to handle the case where we run out of data to send by
pausing the transmission and resuming it once there's more data to send.

This commit extends the `ChunkedRestResponseBody` interface to allow it
to express that it has run out of chunks for immediate transmission, but
that the body will continue at a later point.

DaveCTurner committed:
Pausable chunked HTTP responses

Today we must collect together in memory everything needed to send a
HTTP response before starting to send it to the client, even if we're
using the `chunked` transfer encoding to bound the memory needed for the
final serialization step.

To properly bound the memory usage on the coordinating node we must
instead be able to start sending the response to the client before we
have collected everything needed to finish it. If we do this then we
must be able to handle the case where we run out of data to send by
pausing the transmission and resuming it once there's more data to send.

This commit extends the `ChunkedRestResponseBody` interface to allow it
to express that it has run out of chunks for immediate transmission, but
that the body will continue at a later point.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -236,13 +240,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+
+                @Override
+                public void onFailure(Exception e) {
+                    // TODO tests of this case
```
The one remaining TODO (as written this doesn't work yet)

DaveCTurner replied:
Turns out I can't quite write the test I want to write because of https://github.com/netty/netty/issues/13816, so I've written a somewhat weaker test that exercises the important bits at least.

DaveCTurner committed:
Add test for the failure path

nik9000 commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -43,6 +44,21 @@ public interface ChunkedRestResponseBody {
+     *                 the body of the response, so there's no good ways to handle an exception here. Completing the listener exceptionally
+     *                 will log an error, abort sending the response, and close the HTTP connection.
+     */
+    void getContinuation(ActionListener<ChunkedRestResponseBody> listener);
```
ES|QL's operations use this:
```
    /**
     * An operator can be blocked on some action (e.g. waiting for some resources to become available).
     * If so, it returns a future that completes when the operator becomes unblocked.
     * If the operator is not blocked, this method returns {@link #NOT_BLOCKED} which is an already
     * completed future.
     */
    default SubscribableListener<Void> isBlocked() {
        return NOT_BLOCKED;
    }

    SubscribableListener<Void> NOT_BLOCKED = SubscribableListener.newSucceeded(null);
```

We should probably both do the same thing because we're solving the same fundamental problem. 

Well, I guess we are different in a critical way - yours is called on the transport thread and ES|QL's are called on ESQL's thread pool.

DaveCTurner replied:
I think the main difference is that Netty already has all the machinery needed to handle the continuation, no need to involve a `SubscribableListener`, a plain old `ActionListener` that calls `Channel#writeAndFlush` is enough.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
TODO is done

DaveCTurner commented:
This is ready to review. However, it has no concrete use cases yet so I'm undecided about whether to merge it now or wait until we actually need it for something. Opinions welcome.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Use #104971

ywangd commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -236,13 +240,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+            ActionListener.run(ActionListener.assertOnce(new ActionListener<>() {
+                @Override
+                public void onResponse(ChunkedRestResponseBody continuation) {
+                    channel.writeAndFlush(
+                        new Netty4ChunkedHttpContinuation(writeSequence, continuation, finishingWrite.combiner()),
+                        finishingWrite.onDone() // pass the terminal listener/promise along the line
+                    );
```
I think this means the data is wrapped as a new message to send via the channel which then goes through the channel pipeline (including `Netty4HttpPipeliningHandler`) again and eventually calls into `doWriteChunkedContinuation`?

DaveCTurner replied:
Right.

ywangd reviewed (commented):
I have not read the tests closely yet. The main code change looks fine to me within my range of understanding. 

Could you please confirm that the change should not increase the duration that a channel is held for completing a response? Today, a channel associated to the client request cannot do anything else before the response is fully computed. With this change, we amortize the channel waiting time throughout the response computation so that it is able to send out partial response more quickly without increasing the total time needed for complete processing.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -236,13 +240,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+        final PromiseCombiner combiner = continuation.combiner();
+        assert currentChunkedWrite == null;
+        final var responseBody = continuation.body();
+        currentChunkedWrite = new ChunkedWrite(combiner, promise, responseBody);
+        // NB "writable" means there's space in the downstream ChannelOutboundBuffer, we aren't trying to saturate the physical channel.
+        while (ctx.channel().isWritable()) {
+            if (writeChunk(ctx, combiner, responseBody)) {
+                finishChunkedWrite();
+                return;
+            }
+        }
```
Right.

DaveCTurner commented:
> Could you please confirm that the change should not increase the duration that a channel is held for completing a response? Today, a channel associated to the client request cannot do anything else before the response is fully computed. With this change, we amortize the channel waiting time throughout the response computation so that it is able to send out partial response more quickly without increasing the total time needed for complete processing.

Right. Or at least, this PR doesn't change anything, there are no actions in production which use the pausability yet. But if we made an existing action pausable then it could start sending its response sooner than it does today, but it'd still take the same amount of work to finish computing & sending the response.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Fixup merge

DaveCTurner committed:
Fix test race

ywangd commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4ChunkedHttpContinuation.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.elasticsearch.rest.ChunkedRestResponseBody;
+
+public final class Netty4ChunkedHttpContinuation implements Netty4HttpResponse {
```
I think we don't want this class to be public since it's fully managed within `Netty4HttpPipeliningHandler`. Maybe it could even be an inner class.

DaveCTurner replied:
It can't be a private inner class because we need to mention it when defining the sealed `Netty4HttpResponse` interface. But we can make it package-private, see 708fa524b95.

ywangd replied:
Yeah I was thinking a package private class with a private constructor.

DaveCTurner replied:
> with a private constructor

A package-private constructor right? A fully private constructor wouldn't be accessible where it's needed. Not sure if I'm missing something here.

ywangd replied:
I meant an inner class of `Netty4HttpPipeliningHandler`, e.g.:
```
public class Netty4HttpPipeliningHandler {
    final class Netty4ChunkedHttpContinuation implements Netty4HttpResponse { 
        private Netty4ChunkedHttpContinuation() { }
    }
}

public sealed interface Netty4HttpResponse permits Netty4FullHttpResponse, Netty4ChunkedHttpResponse, Netty4HttpPipeliningHandler.Netty4ChunkedHttpContinuation {
}
```

ywangd commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -198,6 +200,8 @@ private void doWrite(ChannelHandlerContext ctx, Netty4HttpResponse readyResponse
             doWriteFullResponse(ctx, fullResponse, promise);
         } else if (readyResponse instanceof Netty4ChunkedHttpResponse chunkedResponse) {
             doWriteChunkedResponse(ctx, chunkedResponse, promise);
+        } else if (readyResponse instanceof Netty4ChunkedHttpContinuation chunkedContinuation) {
+            doWriteChunkedContinuation(ctx, chunkedContinuation, promise);
```
I wonder whether we need to assert that there was a `Netty4ChunkedHttpResponse` preceeds the continuation. for with the same sequence to ensure we won't get the continuation as the first part of a response?

DaveCTurner replied:
The code structure prevents this I think (i.e. the continuation listener for a particular `writeSequence` is not even created until sending the initial `Netty4ChunkedHttpResponse`, and is wrapped in an `assertOnce`. I added one more assertion in 4954c8e2fdd to help document how this works but I don't see an easy way to do exactly what you suggest.

ywangd replied:
It's likely me overthinking about this. I agree the code itself here is fine. I was thinking that if an API responds with a Continuation first, it could be helpful to catch the issue (of misuse). This comment is related to the other one where I am suggesting reducing the visibilty, also to prevent misuse.

DaveCTurner replied:
This is all hidden in the `transport-netty4` module which means it's already pretty much inaccessible (indeed invisible) to everyone else. I don't think we need more protection than that really.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Tighter visibility

DaveCTurner committed:
Assert continuations received in order

ywangd commented:
If the continuation takes a while and user closes the connection before it completes, I think the error handling is about the same as a full response that comes after channel is closed? If the request is cancellable, does the cancellation propagate to the continuation computation? Maybe yes, if the same task computes the whole response? I guess it's ok even it does not propagate since it can be worked as a separate item. Apologies that I haven't looked into these closely to be sure. I think it's easier to just ask. 😄

DaveCTurner commented:
> If the request is cancellable, does the cancellation propagate to the continuation computation? Maybe yes, if the same task computes the whole response?

This is a good question. It's kind of subtle, this doesn't really integrate with the tasks framework as you might expect:

- `org.elasticsearch.rest.action.RestCancellableNodeClient#doExecute` will unregister the task from channel-close notifications as soon as the client action completes, before it even starts to send the REST response
- `org.elasticsearch.tasks.TaskManager#registerAndExecute` will unregister the task from the task manager as soon as the client action completes so it can't be cancelled from then on either.

If the HTTP connection closes after the client action completes, we notify the caller via `org.elasticsearch.rest.RestResponse#close`, so it's really up to the caller to handle that. In practice I expect most callers would explicitly create a task to represent the computation of the full response and propagate cancellation that way.

DaveCTurner committed:
Comments about cancellation

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Add test for client cancellation

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner commented:
I added the missing test in [2022d7a](https://github.com/elastic/elasticsearch/pull/104851/commits/2022d7a5f424b79a1e2674920cac2d4634014b4e). Turns out to be a good idea, it found a bug :)

DaveCTurner commented:
Actually the bug exists in `main` today too, see https://github.com/elastic/elasticsearch/pull/105258.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

pxsalehi reviewed (commented):
I've used this PR just as an excuse to learn more about this part of the codebase. Don't really have any input for the proposed changes. 

Would it make sense to ask someone with more Netty/Network experience for a review as well?

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Simplify leak detection in tests

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -43,6 +47,30 @@ public interface ChunkedRestResponseBody {
+
+     * <p>Note that the {@link Task} corresponding to any invocation of {@link Client#execute} completes as soon as the client action
+     * returns its response, so it no longer exists when this method is called and cannot be used to receive cancellation notifications.
+     * Instead, if the HTTP channel is closed while sending a response (including while waiting for a continuation) then the REST layer
+     * will invoke {@link RestResponse#close}. Implementations will typically explicitly create a {@link CancellableTask} to represent the
```
> (including while waiting for a continuation)

Hmm on reflection this bit seems implausible. Maybe it only happens the next time we supply a continuation. Needs a test at least.

DaveCTurner replied:
Yep this bit was wrong, adjusted to match reality.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Adjust tests to show cancellation behaviour & fix comment to match

ywangd reviewed (commented):
I think we need update `isDone` usage in the following place
https://github.com/elastic/elasticsearch/blob/4eaff75ed82306f56dba965054b7b72829b81c91/server/src/main/java/org/elasticsearch/rest/RestResponse.java#L88

Also another usage in test similar to how `DefaultRestChannelTests` is updated.
https://github.com/elastic/elasticsearch/blob/4eaff75ed82306f56dba965054b7b72829b81c91/test/framework/src/main/java/org/elasticsearch/rest/RestResponseUtils.java#L40

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Complete listener even on impossible path

DaveCTurner committed:
Assertions to catch invalid usage

DaveCTurner committed:
isDone Javadoc

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -102,6 +131,16 @@ public boolean isDone() {
+
+            @Override
+            public void getContinuation(ActionListener<ChunkedRestResponseBody> listener) {
+                assert false : "no continuations";
```
I'm going for `IllegalStateException` but yes that's fair, see 0b91c7f85835f0fd90b01faef283f511ecdfca45.

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -180,6 +219,16 @@ public boolean isDone() {
+            }
+
+            @Override
+            public void getContinuation(ActionListener<ChunkedRestResponseBody> listener) {
+                assert false : "no continuations";
```
I'm going for `IllegalStateException` but yes that's fair, see 0b91c7f85835f0fd90b01faef283f511ecdfca45.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -236,13 +242,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+            ActionListener.run(ActionListener.assertOnce(new ActionListener<>() {
+                @Override
+                public void onResponse(ChunkedRestResponseBody continuation) {
+                    channel.writeAndFlush(
+                        new Netty4ChunkedHttpContinuation(writeSequence, continuation, finishingWrite.combiner()),
+                        finishingWrite.onDone() // pass the terminal listener/promise along the line
+                    );
```
I don't think so, this code runs on the event loop already so these writes will still either complete or fail properly even if we're in the process of shutting down.

DaveCTurner replied:
Ah yes you're quite right, thanks

DaveCTurner replied:
I've applied a fix in 6fac79de44bf4370837c76442b48e2b404beb5f6 but it had to be a little different from #105306 since we don't have a listener to wrap any more. But then I realised we could do the same thing at the outer layer too, and extract a utility, and in fact what we do today at the outer layer is also a little questionable too. I made the relevant changes in https://github.com/elastic/elasticsearch/pull/105486 and will migrate to that utility once it's available here.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -236,13 +242,54 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+                        e
+                    );
+                    channel.close().addListener(ignored -> {
+                        finishingWrite.combiner().add(channel.newFailedFuture(e));
+                        finishingWrite.combiner().finish(finishingWrite.onDone());
```
I'd rather wait for the whole combiner to complete (and wonder whether we should do that in other places too). And `channel.newFailedFuture()` shouldn't be able to fail so `safeFailPromise` isn't necessary to fail that.

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -43,6 +47,31 @@ public interface ChunkedRestResponseBody {
      */
     boolean isDone();
```
++ see 751ce757272a740b072cb70945a6e006b60165ae.

DaveCTurner commented on modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpClient.java:
```diff
@@ -137,9 +138,20 @@ private synchronized List<FullHttpResponse> sendRequests(final SocketAddress rem
             for (HttpRequest request : requests) {
-                channelFuture.channel().writeAndFlush(request);
+                if (ESTestCase.randomBoolean()) {
+                    channelFuture.channel().writeAndFlush(request);
```
I don't think so here either, this code runs on the event loop already so these writes will still either complete or fail properly even if we're in the process of shutting down.

DaveCTurner replied:
Oh wait sorry I was thinking of somewhere else. We're not on the event loop here indeed. Bear with me...

DaveCTurner replied:
Yes no races here - we're not even waiting for any promises to complete. Instead, we're blocking on the `latch` until we receive all the responses we expect, which certainly implies that everything we queued up has been flushed out over the wire.

DaveCTurner commented:
> I think we need update `isDone` usage in the following place

In fact we can't deal with the case where `isDone()` returns `true` but `isEndOfResponse()` returns `false` here. I guess we could move things around a bit to fix that, but in practice there's no need to support this, because (a) we should really wait until we're ready to send at least some data before even sending the headers, and (b) if we're absolutely sure we want to do this, we could send a body with a single empty chunk. I added an assertion to protect against this in d5e4031ea2f9c1f84a15d9119131152814eda69f.

> Also another usage in test similar to how `DefaultRestChannelTests` is updated.

I'm going to leave that until it's needed (and added an assertion about this in d5e4031ea2f9c1f84a15d9119131152814eda69f too). If we made this util support continuations then we'd have to make it async, or possibly-blocking, and I'd rather do neither. We can fix this up when it's needed.

ywangd commented on modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java:
```diff
@@ -0,0 +1,657 @@
+                }
+            );
+            if (randomBoolean()) {
+                safeSleep(10); // make it more likely the request started executing
```
Nit: maybe `safeSleep(scaledRandomIntBetween(10, 500))`

ywangd commented on modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java:
```diff
@@ -0,0 +1,657 @@
+                        return new RestChannelConsumer() {
+                            @Override
+                            public void close() {
+                                refs.decRef();
```
Nit: maybe `localRefs.decRef()` here and below?

DaveCTurner replied:
It's important that `refs` hasn't been changed here - modified this to use `localRefs` but added an explicit assertion about that.

ywangd reviewed (approved):
LGTM

Thanks for being patient with me for the review process. I learnt a lot from it.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
scaledRandomIntBetween

DaveCTurner committed:
localRefs

DaveCTurner committed:
Fix netty/netty#8007 leak

DaveCTurner commented:
Just an update, there's still a couple more things to do here before this is completely ready:

- Adapt the leak protection added in 6fac79de44bf4370837c76442b48e2b404beb5f6 to use something a bit more like `org.elasticsearch.transport.netty4.Netty4Utils#safeWriteAndFlush`
- Adapt `testContinuationFailure` to the new Netty version and make it check some more things.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Fix up tests

DaveCTurner committed:
Better shutdown protection

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Fixup

henningandersen commented on server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java:
```diff
@@ -39,10 +43,36 @@ public interface ChunkedRestResponseBody {
     Logger logger = LogManager.getLogger(ChunkedRestResponseBody.class);
```
I wonder if we should change the name of this interface to `ChunkedRestResponseBodyPart`? That would make `isDone` less ambiguous (naming-wise). We could also rename `isDone` to `isPartDone` to make this more explicit.

I think the javadoc above could also be improved to mention that it can be multi-part.

DaveCTurner replied:
++ I will do this in a follow-up to avoid introducing too much noise here. I pushed 4570c922b1a to improve the Javadocs here tho.

henningandersen commented on modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java:
```diff
@@ -224,16 +229,75 @@ private void doWriteChunkedResponse(ChannelHandlerContext ctx, Netty4ChunkedHttp
+                    }
+                }
+
+            }), finishingWriteBody::getContinuation);
```
Given that `getContinuation` is expected to be async and take a while, I wonder if we should add some sort of prefetching of these to minimize the stop-go sending here? Clearly something we can do in a follow-up, this is fine as is for now.

DaveCTurner replied:
Yes, but I was expecting to push that responsibility onto the implementation(s) since they will know much more about the costs & benefits of doing so. For instance, we could use a queue to allow parts to arrive asynchronously, and only return `true` from `isDone` if the queue is empty.

DaveCTurner replied:
Also some implementations will be able to handle certain reorderings of parts (e.g. entries in a `.zip` file) but there's no reasonable way to expose that kind of subtlety here IMO.

henningandersen commented on server/src/test/java/org/elasticsearch/http/DefaultRestChannelTests.java:
```diff
@@ -722,27 +788,13 @@ public HttpResponse createResponse(RestStatus status, ChunkedRestResponseBody co
+                    RestResponse.chunked(
+                        RestStatus.OK,
+                        new TestBody(responseBody, between(0, 3)),
+                        () -> assertTrue(isClosed.compareAndSet(false, true))
```
Could we also assert `isDone() && testBody.isEndOfReponse()` (on the test body)?

DaveCTurner replied:
Think so, yes, see 0d923f42bbe.

DaveCTurner replied:
Wait, no, or at least not very easily. We create a sequence of distinct `TestBody` instances, one for each part. I suppose we could capture them on creation but it does rather obscure the point of this test to do so. I'd rather leave these out - I think this is adequately covered elsewhere.

DaveCTurner replied:
Actually maybe not too bad, see a2c385ea17f.

henningandersen reviewed (approved):
LGTM.

Relying a bit on Yang's thorough test review, though i skimmed and it looks like a reasonable set of tests for this.

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses

DaveCTurner committed:
Javadocs

DaveCTurner committed:
Assert body done in test

DaveCTurner committed:
Revert "Assert body done in test"

This reverts commit 0d923f42bbe7ee3e84b702927785094f25d28f3d.

DaveCTurner committed:
Capture parts & assert complete

DaveCTurner committed:
Merge branch 'main' into 2024/01/29/pausable-chunked-responses
